src/Microsoft.FeatureManagement/FeatureFilters/TimeWindowFilter.cs

```diff
@@ -56,7 +69,83 @@ public Task<bool> EvaluateAsync(FeatureFilterEvaluationContext context)
                 return Task.FromResult(false);
             }
 
-            return Task.FromResult((!settings.Start.HasValue || now >= settings.Start.Value) && (!settings.End.HasValue || now < settings.End.Value));
+            //
+            // Hit the first occurrence of the time window
+            if ((!settings.Start.HasValue || now >= settings.Start.Value) && (!settings.End.HasValue || now < settings.End.Value))
+            {
+                return Task.FromResult(true);
+            }
+
+            if (settings.Recurrence != null)
+            {
+                //
+                // The reference of the object will be used for hash key.
+                // If there is no pre-bounded settings attached to the context, there will be no cached filter settings and each call will have a unique settings object.
+                // In this case, the cache for recurrence settings won't work.
+                if (context.Settings == null)
+                {
+                    return Task.FromResult(RecurrenceEvaluator.MatchRecurrence(now, settings));
+                }
+
+                DateTimeOffset cachedTime = _recurrenceCache.GetOrAdd(
```

Comment:
We should be able to have something cleaner to follow where we don't invoke CalculateSurroundingOccurrences twice. Something like:

```
DateTimeOffset cachedtime;

if (!_recurrenceCache.TryGetValue(_settings, out cachedTime))
{
    _recurrenceCache.AddOrUpdate(_settings, GetCacheValue));

    cachedTime = _recurrenceCache[_settings];
}
```

Reply:
Updated as follow:
`RecurrenceEvaluator.CalculateClosestStart`
``` C#
/// <summary>
/// Calculate the start time of the closest active time window.
/// <param name="time">A timestamp.</param>
/// <param name="settings">The settings of time window filter.</param>
/// <returns>The start time of the closest active time window or null if the recurrence range surpasses its end.</returns>
/// </summary>
public static DateTimeOffset? CalculateClosestStart(DateTimeOffset time, TimeWindowFilterSettings settings)
{
    CalculateSurroundingOccurrences(time, settings, out DateTimeOffset? prevOccurrence, out DateTimeOffset? nextOccurrence);

    if (time < settings.Start.Value)
    {
        return nextOccurrence.Value;
    }

    if (prevOccurrence != null)
    {
        bool isWithinPreviousTimeWindow =
                time < prevOccurrence.Value + (settings.End.Value - settings.Start.Value);

        if (isWithinPreviousTimeWindow)
        {
            return prevOccurrence.Value;
        }

        if (nextOccurrence != null)
        {
            return nextOccurrence.Value;
        }
    }

    return null;
}
```

Recurrence evaluation in TimeWIndowFilter
``` C#
DateTimeOffset? closestStart = _recurrenceCache.GetOrAdd(settings, RecurrenceEvaluator.CalculateClosestStart(now, settings));

if (closestStart == null || now < closestStart.Value)
{
    return Task.FromResult(false);
}

if (now < closestStart.Value + (settings.End.Value - settings.Start.Value))
{
    return Task.FromResult(true);
}

closestStart = RecurrenceEvaluator.CalculateClosestStart(now, settings);

_recurrenceCache.AddOrUpdate(
    settings,
    (_) => throw new KeyNotFoundException(),
    (_, _) => closestStart);

if (closestStart == null || now < closestStart.Value)
{
    return Task.FromResult(false);
}

return Task.FromResult(now < closestStart.Value + (settings.End.Value - settings.Start.Value));
```